File: STRUKTUR DATA (SMSTR 2)/materi UAS/multiarytree.py
#MULTIARY TREE

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
class Queue:
    def __init__(self):
        self.front = None
        self.rear = None
    def isEmpty(self):
        return self.front is None
    def enqueue(self, data):
        new_node = Node(data)
        if self.isEmpty():
            self.front = new_node
            self.rear = new_node
        else:
            self.rear.next = new_node
            self.rear = new_node
    def dequeue(self):
        if self.isEmpty():
            print("queue is empty")
            return None
        else:
            dequeue = self.front.data
            if self.front == self.rear:
                self.front = None
                self.rear = None
            else:
                self.front = self.front.next
            return dequeue
    def peek(self):
        if self.isEmpty():
            print("queue is empty")
            return None
        else:
            return self.front.data
    def display(self):
        if self.isEmpty():
            print("queue is empty")
        else:
            current = self.front
            while current:
                print(current.data, end=" ")
                current = current.next
            print()
"""
queue = Queue()
queue.enqueue(10)
queue.enqueue(20)
queue.enqueue(30)
queue.enqueue(40)
print("queue setelah enqueue: ")
queue.display()
"""

queue = Queue()
while True:
    print("\nMenu:\n1.Enqueue\n2.Dequeue\n3.Peek\n4.Display\n5.Exit")
    choice = input("pilih (1/2/3/4/5): ")
    if choice == "1":
        data = input("data yang ingin dimasukkan keantrian: ")
        queue.enqueue(data)
        print("data", data, "telah dimasukan kedalam antrian")
    elif choice == "2":
        removed_data = queue.dequeue()
        if removed_data is not None:
            print("data", removed_data, "telah dihapus dari antrian")
    elif choice == "3":
        front_data = queue.peek()
        if front_data is not None:
            print("elemen pertama dalam antrian: ", front_data)
    elif choice == "4":
        print("isi dari antrian: ")
        queue.display()
    elif choice == "5":
        print("program berhenti")
        break
    else:
        print("pilihan tidka valid. coba lagi")
        

File: STRUKTUR DATA (SMSTR 2)/materi UAS/multiarytree.cpp
#include <iostream>

using namespace std;

class Node {
public:
    int data;
    Node* next;

    Node(int data) : data(data), next(nullptr) {}
};

class Queue {
private:
    Node* front;
    Node* rear;

public:
    Queue() : front(nullptr), rear(nullptr) {}

    bool isEmpty() {
        return front == nullptr;
    }

    void enqueue(int data) {
        Node* new_node = new Node(data);
        if (isEmpty()) {
            front = new_node;
            rear = new_node;
        } else {
            rear->next = new_node;
            rear = new_node;
        }
    }

    int dequeue() {
        if (isEmpty()) {
            cout << "queue is empty" << endl;
            return -1; // or some error code
        } else {
            int dequeue_data = front->data;
            if (front == rear) {
                front = nullptr;
                rear = nullptr;
            } else {
                front = front->next;
            }
            return dequeue_data;
        }
    }

    int peek() {
        if (isEmpty()) {
            cout << "queue is empty" << endl;
            return -1; // or some error code
        } else {
            return front->data;
        }
    }

    void display() {
        if (isEmpty()) {
            cout << "queue is empty" << endl;
        } else {
            Node* current = front;
            while (current != nullptr) {
                cout << current->data << " ";
                current = current->next;
            }
            cout << endl;
        }
    }
};

/*
int main() {
    Queue queue;
    queue.enqueue(10);
    queue.enqueue(20);
    queue.enqueue(30);
    queue.enqueue(40);
    cout << "queue setelah enqueue: " << endl;
    queue.display();
}
*/

int main(){
    Queue queue;
    while (true) {
        cout << "\nMenu:\n1.Enqueue\n2.Dequeue\n3.Peek\n4.Display\n5.Exit" << endl;
        cout << "pilih (1/2/3/4/5): ";
        string choice;
        cin >> choice;

        if (choice == "1") {
            int data;
            cout << "data yang ingin dimasukkan ke antrian: ";
            cin >> data;
            queue.enqueue(data);
            cout << "data " << data << " telah dimasukkan kedalam antrian" << endl;
        } else if (choice == "2") {
            int removed_data = queue.dequeue();
            if (removed_data != -1) {
                cout << "data " << removed_data << " telah dihapus dari antrian" << endl;
            }
        } else if (choice == "3") {
            int front_data = queue.peek();
            if (front_data != -1) {
                cout << "elemen pertama dalam antrian: " << front_data << endl;
            }
        } else if (choice == "4") {
            cout << "isi dari antrian: " << endl;
            queue.display();
        } else if (choice == "5") {
            cout << "program berhenti" << endl;
            break;
        } else {
            cout << "pilihan tidak valid. coba lagi" << endl;
        }
    }

    return 0;
}